kernel/src/scan/mod.rs

```diff
@@ -294,7 +353,197 @@ fn parse_partition_value(raw: Option<&String>, data_type: &DataType) -> DeltaRes
     }
 }
 
-#[cfg(all(test, feature = "default-client"))]
+/// Get the state needed to process a scan. In particular this returns a triple of
+/// (all_fields_in_query, fields_to_read_from_parquet, have_partition_cols) where:
+/// - all_fields_in_query - all fields in the query as [`ColumnType`] enums
+/// - fields_to_read_from_parquet - Which fields should be read from the raw parquet files
+/// - have_partition_cols - boolean indicating if we have partition columns in this query
+fn get_state_info<'a>(
+    schema: &'a Schema,
+    partition_columns: &[String],
+) -> (Vec<ColumnType<'a>>, Vec<StructField>, bool) {
+    let mut have_partition_cols = false;
+    let mut read_fields = Vec::with_capacity(schema.fields.len());
+    // Loop over all selected fields and note if they are columns that will be read from the
+    // parquet file ([`ColumnType::Selected`]) or if they are partition columns and will need to
+    // be filled in by evaluating an expression ([`ColumnType::Partition`])
+    let column_types = schema
+        .fields()
+        .map(|field| {
+            if partition_columns.contains(field.name()) {
+                // todo: this is slow(ish)
+                // Store the raw field, we will turn it into an expression in the inner loop
+                // since the expression could be different for each add file
+                have_partition_cols = true;
+                ColumnType::Partition(Cow::Borrowed(field))
+            } else {
+                // Add to read schema, store field so we can build a `Column` expression later
+                // if needed (i.e. if we have partition columns)
+                read_fields.push(field.clone());
+                ColumnType::Selected(Cow::Borrowed(field))
+            }
+        })
+        .collect();
+    (column_types, read_fields, have_partition_cols)
+}
+
+pub fn selection_vector(
+    engine_interface: &dyn EngineInterface,
+    descriptor: &DeletionVectorDescriptor,
+    table_root: &Url,
+) -> DeltaResult<Vec<bool>> {
+    let fs_client = engine_interface.get_file_system_client();
+    let dv_treemap = descriptor.read(fs_client, table_root)?;
+    Ok(treemap_to_bools(dv_treemap))
+}
+
+pub fn transform_to_logical(
```

Comment:
This method still seems to reflect the old approach, of passing every engine data through kernel, while kernel creates a new expression to apply file-constant data etc for each engine data that it receives.

Have you explored yet the new approach that we've been discussing, where we make file-constant values available to the engine directly, so it can inject them into the engine data? Then we can prepare just one expression at the start of the scan which applies to each engine data in turn?

Reply:
I've exposed the constant values, which is what duckdb will use. Currently in the FFI this is just the partition values, which is can poke at.

Have not had time to work on preparing the single expression approach since we don't need that for duckdb yet. Ideally that would be another PR unless we think it'll massively change the shape of things.